File: problems/algorithms/2116/2116_check_par_string.py
class Solution:
    def canBeValid(self, s: str, locked: str) -> bool:
        n = len(s)
        if (n%2 == 1):
            return False
        left = 0
        v = 0

        for i in range(n):
            if locked[i] == "0":
                v += 1
            elif s[i] == "(":
                left += 1
            elif s[i] == ")":
                if left > 0:
                    left -= 1 
                elif v > 0:
                    v -= 1 
                else:
                    return False

        res = 0 
        for i in range(n-1, -1, -1):
            if locked[i] == "0":
                res -= 1
                v -= 1
            elif s[i] == "(":
                res += 1 
                left -= 1
            elif s[i] == ")":
                res -= 1
            if res > 0:
                return False
            if left == 0 and res == 0:
                break
        
        return left <= 0

File: problems/algorithms/2116/2116_check_par_string.c
#include <stdbool.h>

bool canBeValid(char* s, char* locked) {
    int left = 0;
    int v = 0;
    int size = 0;

    for (int i = 0; s[i] != '\0'; i++) {
        size ++;
        if (locked[i] == '0') {
            v++;
        } else if (s[i] == '(') {
            left++;
        } else if (s[i] == ')') {
            if (left > 0) {
                left--;
            } else if (v > 0) {
                v--;
            } else {
                return false;
            }
        }
    }

    if (size % 2 == 1) {
        return false;
    }
    
    int res = 0;
    for (int i = size - 1; i >= 0; i--) {
        if (locked[i] == '0') {
            res--;
            v--;
        } else if (s[i] == '(') {
            res++;
            left--;
        } else if (s[i] == ')') {
            res--; 
        }
        if (res > 0) {
            return false;
        }
        if (left <= 0 && res == 0) {
            break;
        }
    }
    return left <= 0;
}